### PR TITLE
feat: add blob-stream queue type with remote storage backend

### LINKED ISSUE #1070: Blob (s3) storage for (stream) queues
**Is your feature request related to a problem? Please describe.**
Stream queues can get very long and fill up disks. Would be nice to be able to off-load storage to s3 for "infinite" storage. 

**Describe the solution you'd like**
Make it possible to store stream queues in s3 for "infinite" storage

### PR DESCRIPTION
**Pull request by viktorerlingsson**
Adds a new `blob-stream` queue type (`x-queue-type: "blob-stream"`) that offloads sealed segments to S3-compatible object storage, enabling virtually unlimited retention with bounded local disk usage. Regular `stream` queues are unaffected — both types can coexist in the same instance.

Active (unsealed) segments stay on local disk; once sealed they are uploaded to remote storage and may be evicted locally based on cache size.

Configuration is per-deployment via `[blob-storage]` in `lavinmq.ini` (or matching `--blob-storage-*` CLI flags).

Implements #1070

## Components

- `BlobStorageClient` — minimal S3-compatible client (PUT, GET, multi-object DELETE, LIST) with upload retry.
- `BlobSegmentCache` — local cache of segments fetched from remote storage, size-bounded eviction, read-ahead prefetch for sequential consumers.
- `BlobMessageStore` — extends the stream message store to resolve offsets across remote-only segments, maintains an append-only segment catalog for fast startup (avoids a full bucket LIST), and coordinates async uploads of sealed segments.
- `BlobStream` — new queue class inheriting from `Stream`, always uses `BlobMessageStore`.

## Configuration

```ini
[blob-storage]
region = eu-north-1
endpoint = https://my-bucket.s3.eu-north-1.amazonaws.com
access_key_id = AKIAIOSFODNN7EXAMPLE
secret_access_key = wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
local_segments_per_stream = 50
```

Endpoint must be virtual-hosted-style (bucket in hostname). All keys also available as `--blob-storage-*` CLI flags.

## Testing

Specs cover upload, eviction, offset resolution across remote-only segments, restart/catalog recovery, and overflow edge cases. A test server under `spec/support/` backs the specs — no live endpoint required.

Additionally, a manual integration test suite under `extras/s3-integration/` exercises the feature end-to-end against a real S3-compatible bucket (needs re-run after refactor).

## Dependencies

- Adds `awscr-signer` for AWS SigV4 signing.

## Open items

- [ ] Tested against AWS S3 only; other providers (MinIO, R2, GCS, B2) not yet verified.
- [ ] Low consume throughput with very small segments (<1MB) — see test report for details.
- [x] No upload backpressure: sustained remote outage grows upload queue unboundedly.
- [ ] Consumer offsets are not persisted to remote storage.

### CODE REVIEW

File: src/lavinmq/config.cr [hunk 1]
```diff
@@ -187,6 +195,32 @@ module LavinMQ
...
+        end
+        p.on("--s3-storage-session-token=TOKEN", "S3 token") do |v|
+          @streams_s3_storage_session_token = v
+        end
+        p.on("--s3-storage-endpoint=ENDPOINT", "S3 endpoint") do |v|
+          @streams_s3_storage_endpoint = v
+        end
+        p.on("--s3-storage-local-segments-per-stream=NUMBER", "Number of local segments per stream (default: 50)") do |v|
+          @streams_s3_storage_local_segments_per_stream = v.to_i
+        end
```
Comment (lukas8219) [inline src/lavinmq/config.cr]: Wondering how much we can move the s3 configuration via Streams definitions. I know there are sensitive info like tokens, secrets etc. But would there be any situations where this kind of flexibility is well-suitable?

What if they might want to storage differently between Blue/Green deployments? Or maybe some multi cloud setup?

It might be too overly complex for a first implementation but something to keep in mind

Comment (lukas8219) [reply src/lavinmq/config.cr]: Like `--s3-storage-local-segments-per-stream=NUMBER` - Streams may differ in workload and thus local segments tuning could be specific to stream

**Review by lukas8219 — COMMENTED** [review]
just small comments since i haven't been able to ran it locally yet

File: src/lavinmq/config.cr (hunk 1 above)
Comment (viktorerlingsson) [reply src/lavinmq/config.cr]: > Wondering how much we can move the s3 configuration via Streams definitions. I know there are sensitive info like tokens, secrets etc. But would there be any situations where this kind of flexibility is well-suitable?
> What if they might want to storage differently between Blue/Green deployments? Or maybe some multi cloud setup?
> It might be too overly complex for a first implementation but something to keep in mind
> Like --s3-storage-local-segments-per-stream=NUMBER - Streams may differ in workload and thus local segments tuning could be specific to stream

Yes, we have discussed being able to change some configuration in definitions, but we decided to try to keep it simple in this first version of S3 streams. Some changes between deployments should still be possible with changing config/cli. So we might look into something like that in the future!

**Comment by carlhoerberg** [comment]
Found this: https://github.com/ysbaddaden/xml.cr

File: src/lavinmq/amqp/queue/stream_s3_message_store.cr [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+          if signer = s3_signer
+            signer.sign(request)
+          else
+            raise "No S3 signer found"
+          end
+        end
+        h
+      end
+
+      def s3_signer : Awscr::Signer::Signers::V4?
```
Comment (spuun) [inline src/lavinmq/amqp/queue/stream_s3_message_store.cr]:
```suggestion
      def s3_signer : Awscr::Signer::Signers::V4
```
Looks like it never will be nil. This make it possible to remove the `if signer = s3_signer` check in `#http`.

File: src/lavinmq/amqp/queue/stream_s3_message_store.cr [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+
+      private def load_stats_from_s3_segments(is_long_queue)
+        wg = WaitGroup.new(@s3_segments.size)
+        segments = Array(Tuple(UInt32, MFile)).new(@s3_segments.size)
+        counter = 0
+        running_fibers = 0
+        @s3_segments.each do |seg, s3file|
+          while running_fibers >= MAX_RUNNING_FIBERS
+            Fiber.yield
+          end
```
Comment (spuun) [inline src/lavinmq/amqp/queue/stream_s3_message_store.cr]: What about spawning `[MAX_RUNNING_FIBERS, @s3_segments.size].min` fibers and distribute work through a channel to get rid of this loop?

File: src/lavinmq/amqp/queue/stream_s3_message_store.cr [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+        loop do
+          sleep 50.milliseconds
+          segments_to_download = segments_that_should_be_downloaded
+          next if segments_to_download.empty?
+          segments_to_download.reject! { |seg| @downloading_segments[seg]? || @segments[seg]? || !@s3_segments[seg]? }
+
+          segments_to_download.each do |segment|
+            while @downloading_segments.size >= MAX_RUNNING_FIBERS
+              sleep 10.milliseconds
+            end
```
Comment (spuun) [inline src/lavinmq/amqp/queue/stream_s3_message_store.cr]: Spawn fibers and distribute work through channel?

**Review by spuun — COMMENTED** [review]
Dunno if it would be better or not, but maybe it's good to have global download_segments and monitor_downloads fibers. Not for a first version of s3 queue thought.

File: src/lavinmq/amqp/queue/stream_s3_message_store.cr (hunk 3 above)
Comment (viktorerlingsson) [reply src/lavinmq/amqp/queue/stream_s3_message_store.cr]: Yes, sounds better! Basically just used the fiber pool part of this and no channels for now though

**Comment by viktorerlingsson** [comment]
## S3 integration test report

All 15 end-to-end tests pass as of 2026-04-13 (harness under `extras/s3-integration/`, run against a real AWS S3 bucket in `eu-north-1`).

### Small segments (512KB, 50K messages)

| # | Test | Result |
|---|------|--------|
| 01 | Publish & async S3 upload | PASS — 101 segments uploaded, ~50K msgs/s publish |
| 02 | Restart & durability (no local segments) | PASS — 49995/50000 recovered, ~7.1K msgs/s consume |
| 03 | Consume purely from S3 | PASS — 49995 consumed at ~8–13K msgs/s, local cap respected |
| 04 | 4-consumer benchmark from S3 | PASS — ~800–1700 msgs/s per consumer under contention |
| 05 | Mixed workload (consume old + publish new) | PASS — 69995 consumed, 40 new segments uploaded |
| 06 | Segment catalog upload + recovery | PASS — catalog uploaded on shutdown, restored on restart |

### Large segments (8MB, 1.7M messages)

| # | Test | Result |
|---|------|--------|
| 10 | Large publish & S3 upload | PASS — 214 segments uploaded, ~63K msgs/s publish |
| 11 | Restart with partial local data | PASS — 1.7M survive, S3 fills gaps |
| 12 | Restart with no local data | PASS — 1.7M recovered from S3 alone |
| 13 | Full consume x3 from S3 | PASS — steady 80–130K msgs/s (peaks 370K when cached) |

### Recovery & edge cases

| # | Test | Result |
|---|------|--------|
| 07 | SIGKILL + catalog deletion | PASS — recovers from individual `.meta` files in S3 |
| 08 | Stream deletion (local + S3 cleanup) | PASS — 203 S3 objects deleted via multi-object DELETE, no segfault |
| 09 | `x-max-length` eviction during upload | PASS — regression test for the `open_new_segment` crash fix |
| 14 | 4 concurrent S3-backed streams | PASS — all 4 recover 1.7M each after no-local-data restart |
| 15 | Offset resolution from S3-only segments | PASS — integer, `"first"`, `"last"` all resolve |

Timestamp-based offsets are not exercised by the harness (`lavinmqperf --consumer-args` is JSON-only, no timestamp type). The code path goes through `find_offset_in_segments` just like integer offsets, so integer coverage is representative.

### Known issues

**Low consume throughput with small segments.** Tests 1–5 show only ~8–13K msgs/s consume from S3 with 512KB segments, vs. 80–130K msgs/s with 8MB segments. Each 512KB segment holds ~500 messages so S3 round-trip overhead dominates. Manual testing hits 50–60K msgs/s in similar conditions, so there's likely headroom — prefetch budget / download concurrency / cache coordination worth investigating.

### Architecture concerns worth flagging

1. **Init complexity.** `S3MessageStore#initialize` is a multi-phase sequence (list S3, load catalog, download meta, download last segment, `super`, load stats, fix offsets, start workers, upload remaining, create cache). Ordering is subtle; partial failures leave unclear state. Candidate for an explicit state-machine refactor.
2. **Concurrent mutation of shared hashes.** `@s3_segments`, `@segment_msg_count`, `@segment_first_offset` are touched by upload workers, download workers, the cache, and the main fiber. Only the cache's own dict uses a `Mutex`; the rest relies on cooperative scheduling. Any S3 HTTP call yields — latent race. Not observed in tests but worth auditing.
3. **Inheritance coupling.** `S3MessageStore` subclasses `StreamMessageStore` and overrides ~10 methods. Tight coupling to base-class internals; refactors of `StreamMessageStore` can silently break S3 behavior. Composition would isolate better.
4. **No upload backpressure.** If rotation outpaces upload throughput (slow network, S3 outage), `@upload_queue` and `@failed_uploads` grow unboundedly. Needs a bounded channel or eviction cap.
